### PR TITLE
Improve the low processor mode sleep precision

### LINKED ISSUE #11030: The editor is less responsive when set to "update only changes"
Godot 3.0 master, also 2.1.3
Windows 10 64 bits

I don't know wether this is fixable or not, but on Windows I always noticed a better response time / better framerate when "update always" was checked on top-right. To the point I always checked it when taking videos, because the experience is smooth and reactive. Now I see the option is going to be deprecated and default to "update changes", which finally convinced me to open this issue for the record.

I noticed this in many situations, for example scrolling the script editor, and took a video of it to show the difference: https://github.com/godotengine/godot/issues/10489#issuecomment-325166871

~~@toger5 didn't experience this on Mac, so it might be OS-specific.~~

I just tried to record a video of it in the 3D viewport but I was unable to make the difference stand out much with OBS, because on my laptop it slows down FPS just enough, which is frustrating...
I retried with tweaked settings: http://zylannprods.fr/dl/godot/subtle_framerate_difference.mp4
(but the most measurable example is the script editor).

It feels like the editor reacts and redraws at 30..40 fps / drops frames, instead of 60 when I set "update changes", which is fairly noticeable and impacts usage experience in a global way.
Edit: the difference is also easily seen when resizing docks or using freelook.

## TL;DR:
**The editor lags if not redrawn after a few seconds because the driver puts the graphics card in low-performance mode, and it's by default. It is very slow at waking back up, causing lag, and ruins editing experience.**
It happens only on Windows with nVidia drivers, and AMD appears to have this too, with OpenGL 3. This gets much worse if the scene is complex.
See the measured framerate showing the lag here https://github.com/godotengine/godot/issues/11030#issuecomment-345572705
And explanation for nVidia: https://github.com/godotengine/godot/issues/11030#issuecomment-363450729
This does not happen (or not as badly) with Godot 2 (OpenGL 2), and I also noticed the exact same issue  in Blender 2.8 beta, while reactivity in 2.79 felt much better. (although a bug existed since 2016 https://developer.blender.org/T49758)
A workaround is to change the driver settings to put the application in maximum performance mode, or to run a 60fps application side-by-side. This might however heat your computer much more.
I hope drivers get fixed in the future (it's been years now), or that Vulkan drivers will work better...

### PR DESCRIPTION
**Pull request by Calinou**
**Windows 64-bit editor build for testing:** https://0x0.st/i-iB.zip (link expires in late 2020)

The Low Processor Usage Mode Sleep Usec setting is now used as a FPS limiter rather than a constant sleep duration.

This will increase CPU/GPU usage slightly due to the higher effective FPS, but the increase in overall smoothness is worth it. When used in the editor, it will now really cap at ~145 FPS (as it's supposed to) instead of ~130 FPS.

If both Force Fps and Low Processor Usage Mode settings are enabled in the project settings, only the setting that causes the highest sleep duration will be retained.

@Zylann Please test this :slightly_smiling_face:

This closes #11030.

### CODE REVIEW

**Comment by Calinou** [comment]
@akien-mga Can this be considered for merging for 3.2.2 (as requested [here](https://github.com/godotengine/godot/issues/11030#issuecomment-640964039)), or is it too late now?

File: main/main.cpp [hunk 1]
```diff
@@ -2090,23 +2090,39 @@ bool Main::iteration() {
...
+		// the actual frame time into account.
+		// Due to the high fluctuation of the actual sleep duration, it's not recommended
+		// to use this as a FPS limiter.
+		OS::get_singleton()->delay_usec(frame_delay * 1000);
+	}
+
+	const int target_fps = Engine::get_singleton()->get_target_fps();
+	if ((target_fps > 0 && !Engine::get_singleton()->is_editor_hint()) ||
+			OS::get_singleton()->is_in_low_processor_usage_mode() ||
+			!OS::get_singleton()->can_draw()) {
```
Comment (akien-mga) [inline main/main.cpp:2109]: Isn't `!OS::get_singleton()->can_draw()` falling through the cracks of this refactoring with no `time_step`?

Comment (akien-mga) [reply main/main.cpp:2109]: I'd suggest refactoring for clarity (first indentation level is function scope):
```cpp
uint64_t dynamic_delay = 0;
if (OS::get_singleton()->is_in_low_processor_usage_mode() || !OS::get_singleton()->can_draw()) {
	dynamic_delay = OS::get_singleton()->get_low_processor_usage_mode_sleep_usec();
}
if (target_fps > 0 && !Engine::get_singleton()->is_editor_hint()) {
	dynamic_delay = MAX(dynamic_delay, (uint64_t)(1000000 / target_fps));
}

if (dynamic_delay > 0) {
	target_ticks += dynamic_delay;
	...
}
```

**Comment by akien-mga** [comment]
> Can this be considered for merging in 3.2.2 (as requested here), or is it too late now?

I guess, we can get it tested in 3.2.2 RC 1. Needs a rebase (and possibly further changes, see my review).

**Comment by Calinou** [comment]
Now that 3.2.2RC1 has been released, it's probably too late to integrate it :slightly_frowning_face: 

Also, we'd have to move this PR's base branch to `3.2`, as low-processor usage mode isn't currently working in the `master` branch.

**Comment by akien-mga** [comment]
> Now that 3.2.2RC1 has been released, it's probably too late to merge it for 3.2.2
> Sorry for the inconvenience.

Well if it's rebased soon I can likely include it in 3.2.2 RC 2, unless you think it's too risky.

> Also, we'd have to move this PR's base branch to `3.2`, as low-processor usage mode isn't currently working in the `master` branch.

It should be merged in `master` nevertheless, even if not working, so that the code is there. If the `3.2` code is too different to easily cherry-pick, you can make a dedicated PR for `3.2`.

**Comment by Calinou** [comment]
@akien-mga Rebased :slightly_smiling_face: 

I tested the code by backporting it to the `3.2` branch, it works well. You'll need to replace `!DisplayServer::get_singleton()->window_can_draw()` with `!OS::get_singleton()->can_draw()`.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**